{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import numba\n",
    "import os\n",
    "\n",
    "def yellow(text : object):\n",
    "    return '\\033[33m' + str(text) + '\\033[0m'\n",
    "\n",
    "def green(text : object):\n",
    "    return '\\033[32m' + str(text) + '\\033[0m'\n",
    "\n",
    "def blue(text : object):\n",
    "    return '\\033[34m' + str(text) + '\\033[0m'\n",
    "\n",
    "def red(text : object):\n",
    "    return '\\033[31m' + str(text) + '\\033[0m'\n",
    "\n",
    "def read_genome_file(file_name :str) -> dict[str, int]:\n",
    "    \"\"\"\n",
    "    Reads a genome file and returns a dictionary with the genome name and the length of the genome.\n",
    "    \"\"\"\n",
    "    genome_dict = {}\n",
    "    with open(file_name, 'r') as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            chromo, length = line.split('\\t')\n",
    "            genome_dict[chromo.strip()] = int(length)\n",
    "    return genome_dict\n",
    "\n",
    "class DataSplitter:\n",
    "    \n",
    "    # Hubs contains the bed file as a list of intervals like this:\n",
    "    # [start, end]\n",
    "    # [start, end]\n",
    "    # ...\n",
    "    # [start, end]\n",
    "    hubs : np.ndarray\n",
    "    \n",
    "    # The parent chromosome\n",
    "    chromo: str\n",
    "\n",
    "    def __parse_bed_as_array(self, bed_path : str) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Parses the bed file and returns an array of intervals.\n",
    "        \"\"\"\n",
    "        intervals = []\n",
    "        with open(bed_path, \"r\") as bed_file:\n",
    "            for line in bed_file:\n",
    "                line = line.strip()\n",
    "                if line == \"\":\n",
    "                    continue\n",
    "\n",
    "                # Split the line into a list of values\n",
    "                chr_, start, end, *rest = line.split(\"\\t\")\n",
    "                \n",
    "                # Check if the chromosome is the same as the parent chromosome\n",
    "                if chr_ != self.chromo:\n",
    "                    continue\n",
    "\n",
    "                # Convert the start and end to int\n",
    "                start = int(start)\n",
    "                end = int(end)\n",
    "\n",
    "                # Add the interval to the list\n",
    "                intervals.append((start, end))\n",
    "\n",
    "        return np.array(intervals)\n",
    "\n",
    "    def __init__(self, chromo : str, hubs_path : str, resolution : int):\n",
    "        self.chromo = chromo\n",
    "        self.resolution = resolution\n",
    "        self.hubs = self.__parse_bed_as_array(hubs_path)\n",
    "\n",
    "    @staticmethod\n",
    "    @numba.njit    \n",
    "    def check_single_pair_nb( hubs : np.ndarray, pair : tuple[int, int], resolution : int):\n",
    "        \"\"\"\n",
    "        Checks wether the given pair is a valid pair (belongs inside the hubs).\n",
    "        \"\"\"\n",
    "        for hub in hubs:\n",
    "            contained = True\n",
    "            for p in pair:\n",
    "                \n",
    "                start = p\n",
    "                end = p + resolution\n",
    "\n",
    "                if start < hub[0] or end > hub[1]:\n",
    "                    contained = False\n",
    "                    break\n",
    "\n",
    "            \n",
    "            # If the pair is inside the hub, return True\n",
    "            if contained:\n",
    "                return True\n",
    "\n",
    "        return False\n",
    "\n",
    "    def check_single_pair(self, pair : tuple[int, int]):\n",
    "        return DataSplitter.check_single_pair_nb(self.hubs, pair, self.resolution)\n",
    "\n",
    "    def split_data(self, tsv_path : str, matrix_path : str, out_folder : str):\n",
    "        \"\"\"\n",
    "        Splits the matrix data into two files:\n",
    "        - one containing the pairs that are inside the hubs\n",
    "        - one containing the pairs that are outside the hubs\n",
    "        \"\"\"\n",
    "\n",
    "        # the file paths.\n",
    "        inside_file_path = out_folder + f\"{self.chromo}_inside.tsv\"\n",
    "        outside_file_path = out_folder + f\"{self.chromo}_outside.tsv\"\n",
    "        # create the output folder if it doesn't exist\n",
    "        if not os.path.exists(out_folder):\n",
    "            os.makedirs(out_folder)\n",
    "\n",
    "        # The inside counts\n",
    "        inside_count = 0\n",
    "        outside_count = 0\n",
    "\n",
    "        # The matrices.\n",
    "        matrix = np.load(matrix_path)\n",
    "\n",
    "        # Opening the files.    \n",
    "        with open(inside_file_path, \"w\") as inside_file, open(outside_file_path, \"w\") as outside_file, open(tsv_path, \"r\") as tsv_file:\n",
    "            # Add amn header to the inside and outside files\n",
    "            header = \"bin1\\tbin2\\tzscore\\tpvalue\\n\"\n",
    "            inside_file.write(header)\n",
    "            outside_file.write(header)\n",
    "        \n",
    "            # Looping over the lines.\n",
    "            for line in tsv_file:\n",
    "                line = line.strip()\n",
    "                if line == \"\":\n",
    "                    continue\n",
    "\n",
    "                # Split the line into a list of values\n",
    "                start1, start2, *rest = line.split(\"\\t\")\n",
    "\n",
    "                # Convert the start and end to int\n",
    "                start1 = int(start1)\n",
    "                start2 = int(start2)\n",
    "\n",
    "                # get the index coefficients\n",
    "                i = start1 // self.resolution\n",
    "                j = start2 // self.resolution\n",
    "\n",
    "                # retrieve the p-value from the matrix\n",
    "                pvalue = matrix[i, j]\n",
    "\n",
    "                line = line + \"\\t\" + str(pvalue)\n",
    "\n",
    "                # Check if the pair is inside the hubs\n",
    "                if self.check_single_pair((start1, start2)):\n",
    "                    inside_file.write(line + \"\\n\")\n",
    "                    inside_count += 1\n",
    "                else:\n",
    "                    outside_file.write(line + \"\\n\")\n",
    "                    outside_count += 1\n",
    "        print(f\"Split the data between {green(inside_count)} inside and {red(outside_count)} outside pairs.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "CHROMOSOMES = [\"chr1\", \"chr2\", \"chr3\", \"chr4\", \"chr5\", \"chr6\", \"chr7\", \"chr8\", \"chr9\", \"chr10\", \"chr11\", \"chr12\", \"chr13\", \"chr14\", \"chr15\", \"chr16\", \"chr17\", \"chr18\", \"chr19\", \"chr20\", \"chr21\", \"chr22\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def join_all_files(out_folder : str):\n",
    "    \"\"\"\n",
    "    Joins all the files in the given folder.\n",
    "    \"\"\"\n",
    "    files = os.listdir(out_folder)\n",
    "    \n",
    "    #sort the files by chromosome:\n",
    "    files.sort(key=lambda x: int(x.split(\"_\")[0].replace(\"chr\", \"\")))\n",
    "    inside_file = out_folder + \"inside.tsv\"\n",
    "    outside_file = out_folder + \"outside.tsv\"\n",
    "\n",
    "    with open(inside_file, \"w\") as inside_file, open(outside_file, \"w\") as outside_file:\n",
    "        for file in files:\n",
    "            if file.endswith(\"inside.tsv\"):\n",
    "                with open(out_folder + file, \"r\") as f:\n",
    "                    # skip the header\n",
    "                    next(f)\n",
    "                    for line in f:\n",
    "                        inside_file.write(line)\n",
    "            elif file.endswith(\"outside.tsv\"):\n",
    "                with open(out_folder + file, \"r\") as f:\n",
    "                    #skip the header\n",
    "                    next(f)\n",
    "                    for line in f:\n",
    "                        outside_file.write(line)\n",
    "\n",
    "            print(f\"Joined {file}\")\n",
    "\n",
    "    # Remove the files\n",
    "    for file in files:\n",
    "        os.remove(out_folder + file)\n",
    "    \n",
    "    print(f\"Successfully joined all the files in {out_folder}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on: \u001b[33mchr1\u001b[0m\n",
      "Split the data between \u001b[32m48684\u001b[0m inside and \u001b[31m1262020\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr1\u001b[0m\n",
      "Working on: \u001b[33mchr2\u001b[0m\n",
      "Split the data between \u001b[32m248515\u001b[0m inside and \u001b[31m1171190\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr2\u001b[0m\n",
      "Working on: \u001b[33mchr3\u001b[0m\n",
      "Split the data between \u001b[32m5511\u001b[0m inside and \u001b[31m1078860\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr3\u001b[0m\n",
      "Working on: \u001b[33mchr4\u001b[0m\n",
      "Split the data between \u001b[32m81913\u001b[0m inside and \u001b[31m864145\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr4\u001b[0m\n",
      "Working on: \u001b[33mchr5\u001b[0m\n",
      "Split the data between \u001b[32m13887\u001b[0m inside and \u001b[31m843104\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr5\u001b[0m\n",
      "Working on: \u001b[33mchr6\u001b[0m\n",
      "Split the data between \u001b[32m29203\u001b[0m inside and \u001b[31m811640\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr6\u001b[0m\n",
      "Working on: \u001b[33mchr7\u001b[0m\n",
      "Split the data between \u001b[32m12472\u001b[0m inside and \u001b[31m676068\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr7\u001b[0m\n",
      "Working on: \u001b[33mchr8\u001b[0m\n",
      "Split the data between \u001b[32m9231\u001b[0m inside and \u001b[31m619178\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr8\u001b[0m\n",
      "Working on: \u001b[33mchr9\u001b[0m\n",
      "Split the data between \u001b[32m13985\u001b[0m inside and \u001b[31m364089\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr9\u001b[0m\n",
      "Working on: \u001b[33mchr10\u001b[0m\n",
      "Split the data between \u001b[32m8823\u001b[0m inside and \u001b[31m519297\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr10\u001b[0m\n",
      "Working on: \u001b[33mchr11\u001b[0m\n",
      "Split the data between \u001b[32m4642\u001b[0m inside and \u001b[31m537618\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr11\u001b[0m\n",
      "Working on: \u001b[33mchr12\u001b[0m\n",
      "Split the data between \u001b[32m15382\u001b[0m inside and \u001b[31m525714\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr12\u001b[0m\n",
      "Working on: \u001b[33mchr13\u001b[0m\n",
      "Split the data between \u001b[32m11734\u001b[0m inside and \u001b[31m317772\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr13\u001b[0m\n",
      "Working on: \u001b[33mchr14\u001b[0m\n",
      "Split the data between \u001b[32m5859\u001b[0m inside and \u001b[31m283855\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr14\u001b[0m\n",
      "Working on: \u001b[33mchr15\u001b[0m\n",
      "Split the data between \u001b[32m28773\u001b[0m inside and \u001b[31m202888\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr15\u001b[0m\n",
      "Working on: \u001b[33mchr16\u001b[0m\n",
      "Split the data between \u001b[32m2539\u001b[0m inside and \u001b[31m197887\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr16\u001b[0m\n",
      "Working on: \u001b[33mchr17\u001b[0m\n",
      "Split the data between \u001b[32m3107\u001b[0m inside and \u001b[31m206260\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr17\u001b[0m\n",
      "Working on: \u001b[33mchr18\u001b[0m\n",
      "Split the data between \u001b[32m3377\u001b[0m inside and \u001b[31m217783\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr18\u001b[0m\n",
      "Working on: \u001b[33mchr19\u001b[0m\n",
      "Split the data between \u001b[32m7048\u001b[0m inside and \u001b[31m110257\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr19\u001b[0m\n",
      "Working on: \u001b[33mchr20\u001b[0m\n",
      "Split the data between \u001b[32m1816\u001b[0m inside and \u001b[31m140414\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr20\u001b[0m\n",
      "Working on: \u001b[33mchr21\u001b[0m\n",
      "Split the data between \u001b[32m8189\u001b[0m inside and \u001b[31m37255\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr21\u001b[0m\n",
      "Working on: \u001b[33mchr22\u001b[0m\n",
      "Split the data between \u001b[32m4911\u001b[0m inside and \u001b[31m43520\u001b[0m outside pairs.\n",
      "Finished: \u001b[32mchr22\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "HUBS_PATH = \"../results/hubs/HMEC.CAGE.bed\"\n",
    "\n",
    "for chromo in CHROMOSOMES:\n",
    "    print(\"Working on: \" + yellow(chromo))\n",
    "    # Creating the splitter\n",
    "    splitter = DataSplitter(chromo, HUBS_PATH, resolution=100_000)\n",
    "    \n",
    "    #Splitting the data\n",
    "    splitter.split_data(\n",
    "        matrix_path=f\"../results/ctcf_enrichment/100kb/{chromo}.npy\",\n",
    "        tsv_path=f\"../results/zscores/100kb/{chromo}.txt\",\n",
    "        out_folder=\"../results/hubs_data/100kb/\"\n",
    "    )\n",
    "\n",
    "    print(\"Finished: \" + green(chromo))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Joined chr1_inside.tsv\n",
      "Joined chr1_outside.tsv\n",
      "Joined chr2_inside.tsv\n",
      "Joined chr2_outside.tsv\n",
      "Joined chr3_inside.tsv\n",
      "Joined chr3_outside.tsv\n",
      "Joined chr4_inside.tsv\n",
      "Joined chr4_outside.tsv\n",
      "Joined chr5_inside.tsv\n",
      "Joined chr5_outside.tsv\n",
      "Joined chr6_inside.tsv\n",
      "Joined chr6_outside.tsv\n",
      "Joined chr7_inside.tsv\n",
      "Joined chr7_outside.tsv\n",
      "Joined chr8_inside.tsv\n",
      "Joined chr8_outside.tsv\n",
      "Joined chr9_inside.tsv\n",
      "Joined chr9_outside.tsv\n",
      "Joined chr10_inside.tsv\n",
      "Joined chr10_outside.tsv\n",
      "Joined chr11_inside.tsv\n",
      "Joined chr11_outside.tsv\n",
      "Joined chr12_inside.tsv\n",
      "Joined chr12_outside.tsv\n",
      "Joined chr13_inside.tsv\n",
      "Joined chr13_outside.tsv\n",
      "Joined chr14_inside.tsv\n",
      "Joined chr14_outside.tsv\n",
      "Joined chr15_inside.tsv\n",
      "Joined chr15_outside.tsv\n",
      "Joined chr16_inside.tsv\n",
      "Joined chr16_outside.tsv\n",
      "Joined chr17_inside.tsv\n",
      "Joined chr17_outside.tsv\n",
      "Joined chr18_inside.tsv\n",
      "Joined chr18_outside.tsv\n",
      "Joined chr19_inside.tsv\n",
      "Joined chr19_outside.tsv\n",
      "Joined chr20_inside.tsv\n",
      "Joined chr20_outside.tsv\n",
      "Joined chr21_inside.tsv\n",
      "Joined chr21_outside.tsv\n",
      "Joined chr22_inside.tsv\n",
      "Joined chr22_outside.tsv\n",
      "Successfully joined all the files in ../results/hubs_data/100kb/\n"
     ]
    }
   ],
   "source": [
    "join_all_files(\"../results/hubs_data/100kb/\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.3 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "f6a9dc70ef1d9ac109a6adb4c16e68d584d15afe2d26145e4986863f3d6029bd"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
