{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.cluster_description import Cluster, ClustersDescription\n",
    "import pyranges as pr\n",
    "import numpy as np\n",
    "\n",
    "_GMT_PATH = \"./data/c5.go.bp.v7.5.1.entrez.gmt\"\n",
    "_FANTOM_PATH = \"./data/FANTOM5_CAGE_peak_entrez_gene_tbl.tsv\"\n",
    "\n",
    "# Parses the GMT file returining a dict containing for each \n",
    "# Returns a dictionary mapping each gene entrez id to the ontology group it belongs to.\n",
    "def parse_gmt(path: str) -> dict[str, list[str]]:\n",
    "    # Create a dictionary to store the results\n",
    "    results = {}\n",
    "\n",
    "    # Open the GMT file\n",
    "    with open(path, \"r\") as f:\n",
    "        # For each line in the file\n",
    "        for line in f:\n",
    "            # Split the line into a list of words\n",
    "            words = line.split(\"\\t\")\n",
    "\n",
    "            # The motif name\n",
    "            gene_set = words[0]\n",
    "            link = words[1]\n",
    "            genes = words[2:]\n",
    "            # strip each gene\n",
    "            genes = [gene.strip() for gene in genes]\n",
    "\n",
    "            # Add the gene set to the dictionary\n",
    "            results[gene_set] = [link, genes]\n",
    "\n",
    "    # invert the dictionary\n",
    "    gmt_inverse : dict[str, list[str]] = {}\n",
    "\n",
    "    # For each gene set\n",
    "    for gene_set, (link, genes) in results.items():\n",
    "        for gene in genes:\n",
    "            if gene not in gmt_inverse:\n",
    "                gmt_inverse[gene] = []\n",
    "            gmt_inverse[gene].append(gene_set)\n",
    "        \n",
    "    return gmt_inverse\n",
    "\n",
    "# read the table from the path\n",
    "def parse_fantom_data(path: str) -> dict[str, str]:\n",
    "    # Create a dictionary to store the results\n",
    "    results: dict[str, str] = {}\n",
    "\n",
    "    header_found = False\n",
    "\n",
    "    # Open the file\n",
    "    with open(path, \"r\") as f:\n",
    "        # For each line in the file\n",
    "        for line in f:\n",
    "            # Go to the next line if it starts with a #\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "\n",
    "            # deal with the header\n",
    "            if not header_found:\n",
    "                header_found = True\n",
    "                continue\n",
    "\n",
    "            # Split the line into a list of words\n",
    "            words = line.split(\"\\t\")\n",
    "            cage_id = words[0].strip()\n",
    "            entrez_id = words[1].strip()\n",
    "\n",
    "            # Add the gene to the dictionary\n",
    "            results[cage_id] = entrez_id\n",
    "        \n",
    "    # Return the results\n",
    "    return results\n",
    "\n",
    "# Maps the given entry id to the gene entrez id\n",
    "def overlaps_to_entrez(overlaps: list[str], peak_to_entrez : dict[str, str]) -> list[str]:\n",
    "    entrez_ids = []\n",
    "    for overlap in overlaps:\n",
    "        entrez_ids.append(peak_to_entrez.get(overlap, None))\n",
    "    \n",
    "    # Remove duplicates and NAs\n",
    "    entrez_ids = [entrez_id for entrez_id in entrez_ids if entrez_id not in [\"NA\", None]]\n",
    "    return entrez_ids\n",
    "\n",
    "# Return the ontology group to which each entrez id belongs to.\n",
    "def map_entrez_to_ontology(entrez_ids: list[str], gene_ontology : dict[str, str]) -> list[str]:\n",
    "    ontology_hist = {}\n",
    "    unknown_entrez_ids = []\n",
    "    for entrez_id in entrez_ids:\n",
    "        if entrez_id not in gene_ontology:\n",
    "            unknown_entrez_ids.append(entrez_id)\n",
    "            continue\n",
    "        ontology_groups = gene_ontology[entrez_id]\n",
    "\n",
    "        for ontology_group in ontology_groups:\n",
    "            if ontology_group not in ontology_hist:\n",
    "                ontology_hist[ontology_group] = 0\n",
    "            ontology_hist[ontology_group] += 1\n",
    "    \n",
    "    return ontology_hist, unknown_entrez_ids\n",
    "\n",
    "\n",
    "def get_groups_for_features(features: list[str], peak_to_entrez: dict[str, str], gene_ontology : dict[str, str], remove_duplicates: bool = False) -> dict[str, int]:\n",
    "    # Get the entrez ids\n",
    "    entrez_ids = overlaps_to_entrez(features, peak_to_entrez)\n",
    "\n",
    "    # Removing duplicates\n",
    "    if remove_duplicates:\n",
    "        entrez_ids = list(set(entrez_ids))\n",
    "        \n",
    "    # Get the ontology\n",
    "    groups, _ = map_entrez_to_ontology(entrez_ids, gene_ontology)\n",
    "    \n",
    "    # Return the groups\n",
    "    return groups\n",
    "\n",
    "# Reading the data.\n",
    "peak_to_entrez = parse_fantom_data(_FANTOM_PATH)\n",
    "gene_ontology = parse_gmt(_GMT_PATH)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting the Background Data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Getting the background\n",
    "background = pr.read_bed(f\"../data/features/HMEC/CAGE/features.bed\")\n",
    "\n",
    "# Getting the representation \n",
    "feature_names = background.Name.values.flatten()\n",
    "expected_groups = get_groups_for_features(feature_names, peak_to_entrez, gene_ontology)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading the Observed Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the information about the cluster enriched in both\n",
    "ENRICHED_IN_BOTH_PATH = \"../results/intersected.tsv\"\n",
    "\n",
    "def parse_enrichment_file(path : str):\n",
    "    results = {}\n",
    "\n",
    "    with open(path, \"r\") as f:\n",
    "        for line in f:\n",
    "            words = line.split(\"\\t\")\n",
    "            chromosome = words[0].strip()\n",
    "            clusters = [w.strip() for w in words[1:]]\n",
    "\n",
    "            results[chromosome] = clusters\n",
    "\n",
    "    return results\n",
    "\n",
    "enriched_clusters = parse_enrichment_file(ENRICHED_IN_BOTH_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading the features.\n",
    "features = pr.read_bed(\"../data/features/HMEC/CAGE/features.bed\")\n",
    "\n",
    "observed_groups = {}\n",
    "\n",
    "for chromo, clusters_names in enriched_clusters.items():\n",
    "    # Reading the clusters.\n",
    "    clusters = ClustersDescription(f\"../data/clusters/HMEC/{chromo}_spec_res.json\", chromo)\n",
    "    \n",
    "    # Looping over the cluster (strings)\n",
    "    for cluster in clusters_names:\n",
    "\n",
    "        # Getting the actual cluster object\n",
    "        cluster = clusters[cluster]\n",
    "\n",
    "        # Counting the overlaps\n",
    "        overlaps = cluster.find_overlaps(chromo, features)\n",
    "        overlaps_names = overlaps.Name.values.flatten()\n",
    "        \n",
    "        # Getting the groups\n",
    "        groups = get_groups_for_features(overlaps_names, peak_to_entrez, gene_ontology)\n",
    "\n",
    "        # Adding the groups to the observed groups\n",
    "        for group, count in groups.items():\n",
    "            if group not in observed_groups:\n",
    "                observed_groups[group] = 0\n",
    "            observed_groups[group] += count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7645/7645 [03:03<00:00, 41.61it/s] \n"
     ]
    }
   ],
   "source": [
    "# performing hypergeometric test\n",
    "from scipy.stats import hypergeom\n",
    "from tqdm import tqdm\n",
    "\n",
    "pvalues = {}\n",
    "\n",
    "for group, expected in tqdm(expected_groups.items()):\n",
    "    observed = observed_groups.get(group, 0)\n",
    "    pvalue = hypergeom.sf(observed - 1, len(background), expected, len(background) - expected)\n",
    "    pvalues[group] = [pvalue, observed, expected]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_pvalues : dict[str, float] = sorted(pvalues.items(), key=lambda x: x[1][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "#transform the sorted pvalues into a dataframe\n",
    "import pandas as pd\n",
    "pvalues_df = pd.DataFrame({}, columns=[\"group\", \"pvalue\", \"observed\", \"expected\"])\n",
    "\n",
    "for group, pvalue in sorted_pvalues:\n",
    "    pvalues_df.loc[len(pvalues_df)] = [group, pvalue[0], pvalue[1], pvalue[2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>observed</th>\n",
       "      <th>expected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GOBP_PEPTIDYL_ARGININE_METHYLATION_TO_ASYMMETR...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GOBP_CONSTITUTIVE_SECRETORY_PATHWAY</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GOBP_MENAQUINONE_METABOLIC_PROCESS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GOBP_VITAMIN_K_METABOLIC_PROCESS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GOBP_PROTECTION_FROM_NON_HOMOLOGOUS_END_JOININ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7640</th>\n",
       "      <td>GOBP_VESTIBULOCOCHLEAR_NERVE_FORMATION</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7641</th>\n",
       "      <td>GOBP_POSITIVE_REGULATION_OF_CIRCADIAN_SLEEP_WA...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7642</th>\n",
       "      <td>GOBP_PLASMA_CELL_DIFFERENTIATION</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7643</th>\n",
       "      <td>GOBP_GASTRIC_MOTILITY</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7644</th>\n",
       "      <td>GOBP_REGULATION_OF_ATRIAL_CARDIAC_MUSCLE_CELL_...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7645 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  group  pvalue  observed  \\\n",
       "0     GOBP_PEPTIDYL_ARGININE_METHYLATION_TO_ASYMMETR...     0.0        20   \n",
       "1                   GOBP_CONSTITUTIVE_SECRETORY_PATHWAY     0.0        18   \n",
       "2                    GOBP_MENAQUINONE_METABOLIC_PROCESS     0.0        32   \n",
       "3                      GOBP_VITAMIN_K_METABOLIC_PROCESS     0.0        32   \n",
       "4     GOBP_PROTECTION_FROM_NON_HOMOLOGOUS_END_JOININ...     0.0        36   \n",
       "...                                                 ...     ...       ...   \n",
       "7640             GOBP_VESTIBULOCOCHLEAR_NERVE_FORMATION     1.0         0   \n",
       "7641  GOBP_POSITIVE_REGULATION_OF_CIRCADIAN_SLEEP_WA...     1.0         0   \n",
       "7642                   GOBP_PLASMA_CELL_DIFFERENTIATION     1.0         0   \n",
       "7643                              GOBP_GASTRIC_MOTILITY     1.0         0   \n",
       "7644  GOBP_REGULATION_OF_ATRIAL_CARDIAC_MUSCLE_CELL_...     1.0         0   \n",
       "\n",
       "      expected  \n",
       "0           19  \n",
       "1           16  \n",
       "2           11  \n",
       "3           22  \n",
       "4           30  \n",
       "...        ...  \n",
       "7640         4  \n",
       "7641         1  \n",
       "7642        21  \n",
       "7643         1  \n",
       "7644        16  \n",
       "\n",
       "[7645 rows x 4 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pvalues_df"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "f6a9dc70ef1d9ac109a6adb4c16e68d584d15afe2d26145e4986863f3d6029bd"
  },
  "kernelspec": {
   "display_name": "Python 3.10.3 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
